src/meta-srv/src/region/supervisor.rs

```diff
@@ -688,56 +698,110 @@ impl RegionSupervisor {
-                        region_id, from_peer_id
-                    );
-                    Ok(())
-                }
-                LeaderPeerChanged { .. } => {
-                    self.deregister_failure_detectors(vec![(from_peer_id, region_id)])
-                        .await;
-                    info!(
-                        "Region's leader peer changed, removed failover detector for region: {}, datanode: {}",
-                        region_id, from_peer_id
-                    );
-                    Ok(())
-                }
-                err => Err(err),
-            };
+    async fn do_failover_tasks(
+        &mut self,
+        tasks: Vec<(RegionMigrationProcedureTask, u32)>,
+    ) -> Result<()> {
+        if tasks.is_empty() {
+            warn!("No failover tasks to execute");
+            return Ok(());
+        }
+        let max_count = tasks.iter().map(|(_, count)| *count).max().unwrap_or(1);
+        let region_ids = tasks.iter().map(|(r, _)| r.region_id).collect::<Vec<_>>();
+        let from_peer = tasks[0].0.from_peer.clone();
+        let from_peer_id = from_peer.id;
+        let to_peer = tasks[0].0.to_peer.clone();
+        let to_peer_id = to_peer.id;
+        let timeout = DEFAULT_REGION_MIGRATION_TIMEOUT * max_count;
+        let trigger_reason = RegionMigrationTriggerReason::Failover;
+        let task = RegionMigrationTask {
+            region_ids,
+            from_peer,
+            to_peer,
+            timeout,
+            trigger_reason,
         };
+        let result = self
+            .region_migration_manager
+            .submit_region_migration_task(task)
+            .await?;
+        self.handle_submit_region_migration_task_result(
+            from_peer_id,
+            to_peer_id,
+            timeout,
+            trigger_reason,
+            result,
+        )
+        .await
+    }
+
+    async fn handle_submit_region_migration_task_result(
+        &mut self,
+        from_peer_id: DatanodeId,
+        to_peer_id: DatanodeId,
+        timeout: Duration,
+        trigger_reason: RegionMigrationTriggerReason,
+        result: SubmitRegionMigrationTaskResult,
+    ) -> Result<()> {
+        if !result.migrated.is_empty() {
+            let detecting_regions = result
+                .migrated
+                .iter()
+                .map(|region_id| (from_peer_id, *region_id))
+                .collect::<Vec<_>>();
+            self.deregister_failure_detectors(detecting_regions).await;
+            info!(
+                "Region has been migrated to target peer: {}, removed failover detectors for regions: {:?}",
+                to_peer_id, result.migrated,
+            )
+        }
+        if !result.migrating.is_empty() {
+            info!(
+                "Region is still migrating, skipping failover for regions: {:?}",
+                result.migrating
+            );
```

Comment:
Just logging, not need to `deregister_failure_detectors`?

Reply:
Yes, once the migration is completed, the failure detectors will be removed.